[CSS] Fix shadow of text-decoration/box with currentcolor

mdubet opened the pull request:
#### c71ce85eb4568e261e01fb6b73cff7cce4dce787
<pre>
[CSS] Fix shadow of text-decoration/box with currentcolor
<a href="https://bugs.webkit.org/show_bug.cgi?id=248126">https://bugs.webkit.org/show_bug.cgi?id=248126</a>
rdar://102542182

Reviewed by Antti Koivisto.

&apos;currentcolor&apos; for shadow (text and box) should be resolved at use time to
inherit the cascade properly and not at specified time.

* LayoutTests/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForShadow):
* Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
(WebCore::Display::BoxDecorationPainter::paintBoxShadow const):
* Source/WebCore/display/css/DisplayStyle.cpp:
(WebCore::Display::deepCopy):
(WebCore::Display::Style::Style):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::applyBoxShadowForBackground):
(WebCore::BackgroundPainter::paintBoxShadow):
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
(WebCore::EllipsisBoxPainter::paint):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForeground):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintBackgroundDecorations):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::ShadowApplier::ShadowApplier):
(WebCore::TextPainter::TextPainter):
(WebCore::TextPainter::paintTextWithShadows):
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/rendering/style/ShadowData.h:
(WebCore::ShadowData::ShadowData):
(WebCore::ShadowData::setColor):
(WebCore::ShadowData::color const):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::paintTextWithShadows):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyTextOrBoxShadowValue):

Canonical link: <a href="https://commits.webkit.org/264120@main">https://commits.webkit.org/264120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c30dbc67c8f1842df7ae0f23b8d24a37c696499e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6751 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7152 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8339 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7004 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6912 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9907 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8438 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4887 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13930 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8966 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6679 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5473 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6065 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1598 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6439 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -163,13 +163,21 @@ static inline std::unique_ptr<ShadowData> blendFunc(const ShadowData* from, cons
+
+        auto& renderStyle = context.client->currentStyle();
+        return renderStyle.colorResolvingCurrentColor(color);
+    };
```
Maybe we can pass in the `RenderStyle` from `PropertyWrapperShadow::blend()`?

mdubet replied:
Extracing the `RenderStyle` from `context` in the caller and passing it instead of doing the extraction here ?

graouts replied:
No, passing it down from `PropertyWrapperShadow::blend()` where the from/to `RenderStyle` are readily available.

mdubet replied:
Indeed, those are the correct `RenderStyle`, not the one from `context`. Fixed ✅

anttijk commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -95,7 +95,7 @@ InlineIterator::TextBoxIterator TextBoxPainter<TextBoxPath>::makeIterator() cons
 
 template<typename TextBoxPath>
-void TextBoxPainter<TextBoxPath>::paint()
+void TextBoxPainter<TextBoxPath>::paint(const RenderStyle& renderStyle)
```
Style is already available as a member. I don't see why it is also added as a parameter.

mdubet replied:
No reason indeed, it's fixed.

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -159,17 +159,17 @@ static inline ShadowStyle blendFunc(ShadowStyle from, ShadowStyle to, const CSSP
     ASSERT(from && to);
     ASSERT(from->style() == to->style());
-
+    
```
Stray whitespace.

graouts commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -973,20 +973,20 @@ static inline Ref<CSSPrimitiveValue> adjustLengthForZoom(const Length& length, c
 }
 
-Ref<CSSValue> ComputedStyleExtractor::valueForShadow(const ShadowData* shadow, CSSPropertyID propertyID, const RenderStyle& style, AdjustPixelValuesForComputedStyle adjust)
+Ref<CSSValue> ComputedStyleExtractor::valueForShadow(const ShadowData* shadow, CSSPropertyID propertyID, const RenderStyle& renderStyle, AdjustPixelValuesForComputedStyle adjust)
```
I would keep `style`, at least in this file I don't think we use `renderStyle` much compared to just `style`.

graouts commented on Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
```diff
@@ -1259,6 +1259,16 @@ void BoxDecorationPainter::paintBoxShadow(PaintingContext& paintingContext, Shad
 
     bool hasOpaqueBackground = m_box.style().backgroundColor().isOpaque();
+    
+    auto getResolvedColor = [](const StyleColor& styleColor) {
```
I would drop the `get` and just call this lambda `resolvedColor`.

graouts commented on Source/WebCore/display/css/DisplayStyle.cpp:
```diff
@@ -91,12 +91,12 @@ Style::Style(const RenderStyle& style, const RenderStyle* styleForBackground)
     if (styleForBackground)
         setupBackground(*styleForBackground);
 
+    // FIXME: m_boxShadow should use a custom data structure with a resolved color (aka Color).
```
Could we have a bugzilla bug filed for this and referenced here?

mdubet replied:
Added

graouts commented on Source/WebCore/rendering/TextBoxPainter.h:
```diff
@@ -1,4 +1,5 @@
 /*
+
```
Remove this line.

graouts commented on Source/WebCore/rendering/TextDecorationPainter.cpp:
```diff
@@ -200,7 +200,7 @@ TextDecorationPainter::TextDecorationPainter(GraphicsContext& context, const Fon
 
 // Paint text-shadow, underline, overline
-void TextDecorationPainter::paintBackgroundDecorations(const TextRun& textRun, const BackgroundDecorationGeometry& decorationGeometry, OptionSet<TextDecorationLine> decorationType, const Styles& decorationStyle)
+void TextDecorationPainter::paintBackgroundDecorations(const RenderStyle& renderStyle, const TextRun& textRun, const BackgroundDecorationGeometry& decorationGeometry, OptionSet<TextDecorationLine> decorationType, const Styles& decorationStyle)
```
I would use `style` vs. `renderStyle`.

anttijk commented on Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
```diff
@@ -1259,6 +1259,16 @@ void BoxDecorationPainter::paintBoxShadow(PaintingContext& paintingContext, Shad
+        // FIXME: StyleColor should have been properly resolved already,
+        // but unfortunately it's not enforced in the type system.
+        ASSERT_NOT_REACHED();
+        return Color { };
```
How would you fix this FIXME? Maybe give some information about why this is not happening.

Since the ASSERT_NOT_REACHED says this is not happening you could just have ASSERT(!styleColor.isAbsoluteColor()).

nt1m replied:
I think we should just call `absoluteColor()` at every call site and `ASSERT(styleColor.isAbsoluteColor())` at the beginning of the function, instead of having the `resolveColor()` lambda.

mdubet replied:
It's not always the same StyleColor object which is passed to the lambda.

anttijk commented on Source/WebCore/rendering/TextPainter.cpp:
```diff
@@ -129,7 +129,7 @@ void TextPainter::paintTextOrEmphasisMarks(const FontCascade& font, const TextRu
 }
 
-void TextPainter::paintTextWithShadows(const ShadowData* shadow, const FilterOperations* colorFilter, const FontCascade& font, const TextRun& textRun, const FloatRect& boxRect, const FloatPoint& textOrigin,
+void TextPainter::paintTextWithShadows(const RenderStyle& style, const ShadowData* shadow, const FilterOperations* colorFilter, const FontCascade& font, const TextRun& textRun, const FloatRect& boxRect, const FloatPoint& textOrigin,
```
Wonder if would be better to RenderStyle& to be a constructor parameter. These argument lists are pretty long.

mdubet replied:
Changed to be a constructor parameter. ✔️

nt1m commented on Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
```diff
@@ -1260,7 +1260,19 @@ void BoxDecorationPainter::paintBoxShadow(PaintingContext& paintingContext, Shad
+        // https://bugs.webkit.org/show_bug.cgi?id=248467
+        ASSERT_NOT_REACHED();
+        return Color { };
+    };
```
Can we confidently do this instead?

```suggestion
    auto resolveColor = [] (const StyleColor& styleColor) {
        // FIXME: StyleColor should have been properly resolved already,
        // but unfortunately it's not enforced in the type system.
        // We should create a new class such as ResolvedShadowData to make it explicit.
        // https://bugs.webkit.org/show_bug.cgi?id=248467
        ASSERT(styleColor.isAbsoluteColor());
        return styleColor.absoluteColor();
    };
```

mdubet replied:
In release build, this would be UB whereas the `Color { }` is not ; so it's kind of a safety net.

nt1m replied:
This isn't shipping code, but an experiment, so we shouldn't care about release builds too much. I'd rather have it crash so we catch issues with the design.

mdubet replied:
Changed

mdubet committed:
[CSS] Fix shadow of text-decoration/box with currentcolor
https://bugs.webkit.org/show_bug.cgi?id=248126
rdar://102542182

Reviewed by Antti Koivisto.

'currentcolor' for shadow (text and box) should be resolved at use time to
inherit the cascade properly and not at specified time.

* LayoutTests/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForShadow):
* Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
(WebCore::Display::BoxDecorationPainter::paintBoxShadow const):
* Source/WebCore/display/css/DisplayStyle.cpp:
(WebCore::Display::deepCopy):
(WebCore::Display::Style::Style):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::applyBoxShadowForBackground):
(WebCore::BackgroundPainter::paintBoxShadow):
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
(WebCore::EllipsisBoxPainter::paint):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintForeground):
(WebCore::TextBoxPainter<TextBoxPath>::paintBackgroundDecorations):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::ShadowApplier::ShadowApplier):
(WebCore::TextPainter::TextPainter):
(WebCore::TextPainter::paintTextWithShadows):
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/rendering/style/ShadowData.h:
(WebCore::ShadowData::ShadowData):
(WebCore::ShadowData::setColor):
(WebCore::ShadowData::color const):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::paintTextWithShadows):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyTextOrBoxShadowValue):

Canonical link: https://commits.webkit.org/264120@main
